### ISSUE TITLE
AGIC Sets AG Rules Incorrectly When Ingres Path is not /

### ISSUE THREAD

**Issue by kylelee24** [Azure/application-gateway-kubernetes-ingress#1013]
**Describe the bug**
When deploying an ingress rule with a path other than `/` (E.g., `/api`), the AGIC (Application Gateway Ingress Controller) configures rules on the AG (Application Gateway) which sets the backend target / HTTP setting as `defaultaddresspool` / `defaulthttpsetting`. This results in a 502.

**To Reproduce**

- Deploy AGIC
- Deploy nginx container with an ingress rule setting path to `/`
- Validate everything is routing correctly where I can access `/` or `/api/` or `/api/health`
- Update ingress rule setting path to `/api`
- Results in 502 on the browser with the AG rules set incorrectly where the backend target / HTTP setting as `defaultaddresspool` / `defaulthttpsetting`

**Ingress Controller details**
* Output of `kubectl describe pod <ingress controller`> . The <ingress controller> pod name can be obtained by running `helm list`. 

```
Name:         ingress-azure-588859d7d-kjh6c
Namespace:    default
Priority:     0
Node:         aks-default-34713689-vmss000000/10.11.0.4
Start Time:   Wed, 09 Sep 2020 17:43:16 -0400
Labels:       app=ingress-azure
              pod-template-hash=588859d7d
              release=ingress-azure
Annotations:  checksum/config: b6a74679883b8fafff0ec12fed2f6271bcaa3d4aa3ae645640486b45091cc0d8
              prometheus.io/port: 8123
              prometheus.io/scrape: true
Status:       Running
IP:           10.11.0.18
IPs:
  IP:           10.11.0.18
Controlled By:  ReplicaSet/ingress-azure-588859d7d
Containers:
  ingress-azure:
    Container ID:   docker://c565adf68611ad67c63cd7bf248c62cf88603896f81c43d513e0bc21918ede84
    Image:          mcr.microsoft.com/azure-application-gateway/kubernetes-ingress:1.2.0-rc3
    Image ID:       docker-pullable://mcr.microsoft.com/azure-application-gateway/kubernetes-ingress@sha256:9331411a7f29cfd49b6d93e045ea6dea44cab07a2b4ce7d6a6448b3eb23d5200
    Port:           <none>
    Host Port:      <none>
    State:          Running
      Started:      Wed, 09 Sep 2020 17:43:42 -0400
    Ready:          True
    Restart Count:  0
    Liveness:       http-get http://:8123/health/alive delay=15s timeout=1s period=20s #success=1 #failure=3
    Readiness:      http-get http://:8123/health/ready delay=5s timeout=1s period=10s #success=1 #failure=3
    Environment Variables from:
      ingress-azure  ConfigMap  Optional: false
    Environment:
      AZURE_CLOUD_PROVIDER_LOCATION:  /etc/appgw/azure.json
      AGIC_POD_NAME:                  ingress-azure-588859d7d-kjh6c (v1:metadata.name)
      AGIC_POD_NAMESPACE:             default (v1:metadata.namespace)
      AZURE_AUTH_LOCATION:            /etc/Azure/Networking-AppGW/auth/armAuth.json
    Mounts:
      /etc/Azure/Networking-AppGW/auth from networking-appgw-k8s-azure-service-principal-mount (ro)
      /etc/appgw/azure.json from azure (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from ingress-azure-token-26448 (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  azure:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/kubernetes/azure.json
    HostPathType:  File
  networking-appgw-k8s-azure-service-principal-mount:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  networking-appgw-k8s-azure-service-principal
    Optional:    false
  ingress-azure-token-26448:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  ingress-azure-token-26448
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason                     Age                 From                                                      Message
  ----     ------                     ----                ----                                                      -------
  Warning  FailedApplyingAppGwConfig  11m (x6 over 6d7h)  azure/application-gateway, ingress-azure-588859d7d-kjh6c  Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded
```

* Output of `kubectl logs <ingress controller>. 

```
I0917 15:22:02.750699       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0917 15:22:02.758253       1 mutate_app_gateway.go:177] BEGIN AppGateway deployment
I0917 15:37:03.392732       1 mutate_app_gateway.go:183] END AppGateway deployment
E0917 15:37:03.392877       1 controller.go:141] Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded
E0917 15:37:03.392973       1 worker.go:62] Error processing event.Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded
I0917 15:37:09.511123       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0917 15:37:09.520090       1 mutate_app_gateway.go:177] BEGIN AppGateway deployment
I0917 15:52:10.199412       1 mutate_app_gateway.go:183] END AppGateway deployment
E0917 15:52:10.199480       1 controller.go:141] Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded
E0917 15:52:10.199648       1 worker.go:62] Error processing event.Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded
I0917 15:52:16.277488       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0917 15:52:16.284813       1 mutate_app_gateway.go:177] BEGIN AppGateway deployment
I0917 16:07:16.976324       1 mutate_app_gateway.go:183] END AppGateway deployment
E0917 16:07:16.976391       1 controller.go:141] Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded
E0917 16:07:16.976403       1 worker.go:62] Error processing event.Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded
I0917 23:04:51.277313       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0917 23:04:51.284213       1 mutate_app_gateway.go:177] BEGIN AppGateway deployment
I0917 23:05:22.115132       1 mutate_app_gateway.go:185] Applied generated Application Gateway configuration
I0917 23:05:22.115207       1 mutate_app_gateway.go:200] cache: Updated with latest applied config.
I0917 23:05:22.116272       1 mutate_app_gateway.go:204] END AppGateway deployment
I0917 23:05:22.116312       1 controller.go:151] Completed last event loop run in: 31.104207149s
I0917 23:05:23.179622       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0917 23:05:23.188097       1 mutate_app_gateway.go:177] BEGIN AppGateway deployment
I0917 23:05:53.884503       1 mutate_app_gateway.go:185] Applied generated Application Gateway configuration
I0917 23:05:53.884526       1 mutate_app_gateway.go:200] cache: Updated with latest applied config.
I0917 23:05:53.885556       1 mutate_app_gateway.go:204] END AppGateway deployment
I0917 23:05:53.885601       1 controller.go:151] Completed last event loop run in: 30.769080862s
W0918 21:09:23.853369       1 reflector.go:402] pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: watch of *v1.Service ended with: very short watch: pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: Unexpected watch close - watch lasted less than a second and no items received
W0918 21:09:23.853704       1 reflector.go:402] pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: watch of *v1.Secret ended with: very short watch: pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: Unexpected watch close - watch lasted less than a second and no items received
W0918 21:09:23.853884       1 reflector.go:402] pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: watch of *v1.Endpoints ended with: very short watch: pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: Unexpected watch close - watch lasted less than a second and no items received
E0918 21:09:23.855965       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: Failed to list *v1.Secret: Get "https://172.16.0.1:443/api/v1/secrets?resourceVersion=2720416": dial tcp 172.16.0.1:443: connect: connection refused
E0918 21:09:23.856085       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: Failed to list *v1.Endpoints: Get "https://172.16.0.1:443/api/v1/endpoints?resourceVersion=2893071": dial tcp 172.16.0.1:443: connect: connection refused
E0918 21:09:23.856149       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: Failed to list *v1.Service: Get "https://172.16.0.1:443/api/v1/services?resourceVersion=2328598": dial tcp 172.16.0.1:443: connect: connection refused
E0918 21:09:23.863251       1 reflector.go:380] pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: Failed to watch *v1beta1.Ingress: Get "https://172.16.0.1:443/apis/extensions/v1beta1/ingresses?allowWatchBookmarks=true&resourceVersion=2328600&timeout=6m45s&timeoutSeconds=405&watch=true": dial tcp 172.16.0.1:443: connect: connection refused
E0918 21:09:23.863333       1 reflector.go:380] pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: Failed to watch *v1.Pod: Get "https://172.16.0.1:443/api/v1/pods?allowWatchBookmarks=true&resourceVersion=2720436&timeout=7m19s&timeoutSeconds=439&watch=true": dial tcp 172.16.0.1:443: connect: connection refused
E0918 21:09:24.865274       1 reflector.go:380] pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: Failed to watch *v1beta1.Ingress: Get "https://172.16.0.1:443/apis/extensions/v1beta1/ingresses?allowWatchBookmarks=true&resourceVersion=2328600&timeout=6m53s&timeoutSeconds=413&watch=true": dial tcp 172.16.0.1:443: connect: connection refused
E0918 21:09:24.865559       1 reflector.go:380] pkg/mod/k8s.io/client-go@v0.0.0-20200326020446-6240434e1ad6/tools/cache/reflector.go:125: Failed to watch *v1.Pod: Get "https://172.16.0.1:443/api/v1/pods?allowWatchBookmarks=true&resourceVersion=2720436&timeout=7m2s&timeoutSeconds=422&watch=true": dial tcp 172.16.0.1:443: connect: connection refused
I0918 21:09:27.321175       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0918 21:09:27.326606       1 mutate_app_gateway.go:164] cache: Config has NOT changed! No need to connect to ARM.
I0918 21:09:27.326622       1 controller.go:151] Completed last event loop run in: 520.837287ms
I0923 21:12:25.191025       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0923 21:12:25.196572       1 mutate_app_gateway.go:164] cache: Config has NOT changed! No need to connect to ARM.
I0923 21:12:25.196595       1 controller.go:151] Completed last event loop run in: 295.852042ms
I0923 21:12:26.242234       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0923 21:12:26.248056       1 mutate_app_gateway.go:164] cache: Config has NOT changed! No need to connect to ARM.
I0923 21:12:26.248103       1 controller.go:151] Completed last event loop run in: 51.307312ms
I0923 21:12:29.329649       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0923 21:12:29.338665       1 mutate_app_gateway.go:177] BEGIN AppGateway deployment
I0923 21:18:40.643706       1 mutate_app_gateway.go:185] Applied generated Application Gateway configuration
I0923 21:18:40.643728       1 mutate_app_gateway.go:200] cache: Updated with latest applied config.
I0923 21:18:40.651378       1 mutate_app_gateway.go:204] END AppGateway deployment
I0923 21:18:40.651405       1 controller.go:151] Completed last event loop run in: 6m11.373433286s
I0923 21:18:41.693066       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0923 21:18:41.700723       1 mutate_app_gateway.go:177] BEGIN AppGateway deployment
I0923 21:33:42.592108       1 mutate_app_gateway.go:183] END AppGateway deployment
E0923 21:33:42.592164       1 controller.go:141] Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded
E0923 21:33:42.592173       1 worker.go:62] Error processing event.Future#WaitForCompletion: context has been cancelled: StatusCode=200 -- Original Error: context deadline exceeded
I0923 21:33:48.667087       1 redirects.go:44] Created redirection configuration sslr-fl-3c2d0720c26570ce121499b4720f5432 for ([nginx.dev.mydomain.org    ],443); not yet linked to a routing rule
I0923 21:33:48.674219       1 mutate_app_gateway.go:177] BEGIN AppGateway deployment
```

* Any Azure support tickets associated with this issue.

**Comment by vishal8k** [comment]
@kylelee24  Try setting the ingress path as `/api/*` Also see whether you have any backend path redirection annotation added to your ingress yaml, If possible post your ingress yaml here for easy debugging.

**Comment by kylelee24** [comment]
@vishal8k, sorry I did have the backend path direction set at the time where I provided the logs however with or without it set resulted in the same behaviour.

What did fix the issue is setting the ingress path as `/api/*`. 

Note: In the rules, the Backend target is set to `defaultadddresspool` and the HTTP setting is set to `defaulthttpsetting` however under Path-based routing, the Target name, HTTP setting name, and Backend pool is correct reflecting the correct generated resources from the AGIC.

![Screenshot 2020-09-24 092153](https://user-images.githubusercontent.com/41482534/94150651-6b9ff680-fe47-11ea-833c-4efba15da050.png)
